### ISSUE TITLE
[06-fused-attention] Determine if FP8 operand B can use 2d block load

### ISSUE THREAD

**Issue by whitneywhtsang** [intel/intel-xpu-backend-for-triton#3572]

**Comment by chengjunlu** [comment]
There is hardware limitation that the block IO load with transpose only can be used for d32 element type.
In the FP8 on Xe3 case, we can use the transpose load as required for VNNI format.

In this FP8 on Xe2 case, we want to load the operands as the layout required by the DPAS with OpsPerChannel=2. Then we can do element-wise upcast FP8 to FP16. Finally use the FP16 DPAS instruction to compute the matrix multiply computation without shuffling data for better performance.

The matrix B is distributed in register of a single DPAS instruction loaded by block IO without transform and transpose is like this:

![Image](https://github.com/user-attachments/assets/0fd310fa-4fec-49b0-98aa-f53d22353bf4)

The number in the matrix is the index of the elements to the base register.

The IGC vectorizer will allocate the elements to each lane. Let's take the threads_per_warp=16 for example. The value distribution is like this:

![Image](https://github.com/user-attachments/assets/a3ecfeff-fc87-4d6a-8288-92643eb14d5f)

But the required layout to the DPSA instruction is like this:

![Image](https://github.com/user-attachments/assets/750f20b1-6a34-4861-9587-12a1879874b5)


We need a low cost in-register transposing to this case..

**Comment by chengjunlu** [comment]
Add a case to debug the transposing in block IO load.

```Python
import torch
import triton
import tempfile


def test_2d_block_io():

    ir = """
#dpas = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 2, threadsPerWarp = 16, warpsPerCTA = [1, 1], repCluster = [1, 1], A = [8, 16], B = [16, 16], C = [8, 16]}>
#dot1 = #ttg.dot_op<{opIdx = 1, parent = #dpas, kWidth=2}>
module attributes {"ttg.num-warps" = 1 : i32, "ttg.threads-per-warp" = 16 : i32} {
  tt.func public @test(%arg0: !tt.ptr<i8>, %height: i64, %width: i64, %stride: i64) {
    %c0_i32 = arith.constant 0 : i32
    %c1_i64 = arith.constant 1 : i64
    %ptrX = tt.make_tensor_ptr %arg0, [%height, %width], [%c1_i64, %stride], [%c0_i32, %c0_i32] {order = array<i32: 1, 0>} : <tensor<64x64xi8, #dot1>>
    %X = tt.load %ptrX {boundaryCheck = array<i32: 0, 1>, ttig.block_io = "column_major"} : !tt.ptr<tensor<64x64xi8, #dot1>>
    tt.print " i8 tensor: " {hex = false, isSigned = array<i32: 0>} : %X : tensor<64x64xi8, #dot1>
    tt.return
  }
}

    """

    with tempfile.NamedTemporaryFile(mode='w', suffix='.ttgir') as f:
        f.write(ir)
        f.flush()
        kernel = triton.compile(f.name)

    M, N = 64, 64
    k = torch.arange(M * N).view(M, N)
    k = k.to(torch.uint8).to("xpu").transpose(0, 1).contiguous().transpose(0, 1)
    print(k.cpu())

    kernel[(1, 1, 1)](k, M, N, k.stride()[1])


if __name__ == "__main__":
    test_2d_block_io()
```

**Comment by chengjunlu** [comment]
@whitneywhtsang 
It seems there is no simple solution for low cost in place transposing so far.

Can we close this issue?

**Comment by whitneywhtsang** [comment]
> [@whitneywhtsang](https://github.com/whitneywhtsang) It seems there is no simple solution for low cost in place transposing so far.
> 
> Can we close this issue?

As discussed offline, more investigation is needed on layout of B with transform.

**Comment by alexbaden** [comment]
Can you summarize the offline discussion?

**Comment by whitneywhtsang** [comment]
Is is possible to manually modify the code to get a performance estimation?

**Comment by etiotto** [comment]
@chengjunlu this issue is in review, what is the associated PR ?

**Comment by chengjunlu** [comment]
Move it to in progress.

**Comment by chengjunlu** [comment]
I have created an experimental branch to verify the load the column major B matrix. (It should work for A as well.)

The code is ready only if:
1. Make sure it is good for performance in GEMM and attention kernel.
2. No performance degression.
3. Need to refactor the block io load lowering at first to reduce the complication in the code.

**Comment by etiotto** [comment]
@chengjun can you please summarize what is currently supported here.

**Comment by chengjunlu** [comment]
> [@chengjun](https://github.com/chengjun) can you please summarize what is currently supported here.

> The 2d block io only can transpose the matrix of i32 type when load matrix from memory to register. To transpose the matrix of type bits number < 32, we need to further transpose matrix inside the register.
> 
> The steps the to load matrix with transposing with 2D block IO:
> 
>  - To load the matrix as d32 type matrix from memory with transposed to register.
>  - (Optional if scalar type < 32 bits) To transpose the MxNxd32 to Mx(32/m)xNxdm inside the register.
> Right now we only use the bitcast operation to transpose the matrix of which the width is equal to the threads per warp for step 2.

I will take a example to explain it:
```
#mma = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 4, threadsPerWarp = 16, warpsPerCTA = [1, 4], repCluster = [1, 2], A = [8, 32], B = [32, 32], C = [8, 32]}>
%14 = tt.load %13 {ttig.block_io = "column_major"} : tensor<128x128x!tt.ptr<i8>, #mma>
```

Load a column major matrix into the register as the layout `#mma` with threadsPerWarp=16 is supported by using 2D block IO. 
After using a 2D block IO load at i32 granularity, we will further do the transposing with bitcast operation. As shown in the LIT case.
```
// CHECK: %[[LOADED:.*]] = triton_gen.2Dblockload {{.*}},  {{.*}},  {{.*}},  {{.*}},  {{.*}},  {{.*}} {elem_size_in_bits = 32, tile_width = 8, tile_height = 16, v_blocks = 1, transpose = true, vnni_transform = false, cache_control = Default} : (!llvm.ptr<1>, i32, i32, i32, i32, i32) -> vector<8xi32>
// CHECK: %[[PACKED_1ST_HALF:.*]] = llvm.shufflevector %[[LOADED]], %[[LOADED]] [0, 1] : vector<8xi32>
// CHECK: llvm.bitcast %[[PACKED_1ST_HALF]] : vector<2xi32> to vector<8xi8>
```

But for the same load operation but a little different `#mma` of which the threadsPerWarp=32.
```
#mma = #ttig.dpas<{repeatCount = 8, systolicDepth = 8, executionSize = 16, opsPerChan = 4, threadsPerWarp = 32, warpsPerCTA = [1, 4], repCluster = [1, 2], A = [8, 32], B = [32, 32], C = [8, 32]}>
%14 = tt.load %13 {ttig.block_io = "column_major"} : tensor<128x128x!tt.ptr<i8>, #mma>
```

We cannot use the 2D load block IO becuase there is no intrinsic could be used to do the register transposing of tile width=16.

I will use inline VISA to enhance this case further.

Eventually, it is expected to load column major matrix of which the element type bits size < 32 with 2D block IO.

**Comment by chengjunlu** [comment]
> Right now we only use the bitcast operation to transpose the matrix of which the width is equal to the threads per warp for step 2.

The draft PR can support to load column major matrix for elements type bits size < 32 where the bitcast op couldn't be used.
https://github.com/intel/intel-xpu-backend-for-triton/pull/5708

**Comment by chengjunlu** [comment]
The transposing 2D load is used for both K and V matrix for 06 turtorial. after the PR https://github.com/intel/intel-xpu-backend-for-triton/pull/5500.

But the performance is not good as using the scatter load because there are more registers spilling with BLOCK IO.  

This issue is fnished. I will create an new issue to track the performance degration on 06 tutorial of FP8.
